### PR TITLE
accessing method visibilities

### PR DESCRIPTION
**Pull request by nobu**
[Feature #12655](https://bugs.ruby-lang.org/issues/12655)

### CODE REVIEW

File: class.c [hunk 1]
```diff
@@ -1732,6 +1771,46 @@ rb_singleton_class(VALUE obj)
     return klass;
 }
 
+/*
+ *  call-seq:
+ *     class.singleton_instance -> object or nil
+ *
+ *  Returns the singelton instance of <i>class</i>, or <code>nil</code>.
```
Comment (mathieujobin) [inline class.c]: small type in singleton

File: internal.h [hunk 2]
```diff
@@ -1480,6 +1480,7 @@ VALUE rb_obj_methods(int argc, const VALUE *argv, VALUE obj);
 VALUE rb_obj_protected_methods(int argc, const VALUE *argv, VALUE obj);
 VALUE rb_obj_private_methods(int argc, const VALUE *argv, VALUE obj);
 VALUE rb_obj_public_methods(int argc, const VALUE *argv, VALUE obj);
+VALUE rb_obj_undefined_methods(int argc, const VALUE *argv, VALUE obj);
```
Comment (mathieujobin) [inline internal.h]: wasn't this moved to `internal/class.h` ? 

also, what happen to this other function that was present in the original patch?

`int rb_obj_basic_to_s_p(VALUE);`

**Comment by mathieujobin** [comment]
will this make it for Ruby 3.1 for Xmas 2021?

**Comment by mathieujobin** [comment]
was this replaced by something else?

**Comment by hsbt** [comment]
This PR pertly merged at https://bugs.ruby-lang.org/projects/ruby-master/repository/git/revisions/7cda7fbbdc14f4262afaa94cdeb5a5987f1eb01a by @jeremyevans

**Pull request closed without merging**